include/libtorrent/io.hpp

```diff
@@ -80,11 +81,16 @@ namespace libtorrent
 				|| std::is_enum<In>::value>::type>
 		inline void write_impl(In data, OutIt& start)
 		{
+			using OutItT = typename std::iterator_traits<OutIt>::value_type;
+			using Byte = typename std::conditional<
+				std::is_same<OutItT, void>::value, char, OutItT>::type;
```

Comment:
I suppose this could be improved in next versions of C++ (or with the use of a new `byte` type), but for now, this is to allow the case when `std::back_insert_iterator<std::string>` is used, in C++11 `::value_type` is `void`. See https://github.com/arvidn/libtorrent/blob/master/src/bt_peer_connection.cpp#L2215
```cpp
std::string remote_address;
std::back_insert_iterator<std::string> out(remote_address);
detail::write_address(remote().address(), out);
```